test/e2e/remediation_test.go

```diff
@@ -99,85 +94,64 @@ func remediation() {
 	Expect(bootstrapClient.Delete(ctx, &workerMachine)).To(Succeed(), "Failed to delete worker Machine")
 
 	Logf("Waiting for worker BMH to be in Available state")
-	Eventually(func(g Gomega) error {
+	Eventually(func(g Gomega) {
 		g.Expect(bootstrapClient.Get(ctx, client.ObjectKey{Namespace: namespace, Name: workerBmh.Name}, &workerBmh)).To(Succeed())
 		g.Expect(workerBmh.Status.Provisioning.State).To(Equal(bmov1alpha1.StateAvailable))
-		return nil
 	}, e2eConfig.GetIntervals(specName, "wait-machine-remediation")...).Should(Succeed())
 
 	Logf("Waiting for 2 BMHs to be Provisioned")
-	Eventually(func(g Gomega) error {
+	Eventually(func(g Gomega) {
 		bmhs, err := getAllBmhs(ctx, bootstrapClient, namespace, specName)
-		if err != nil {
-			Logf("Error: %v", err)
-			return err
-		}
+		g.Expect(err).NotTo(HaveOccurred())
 		provisioningBMHs := filterBmhsByProvisioningState(bmhs, bmov1alpha1.StateProvisioned)
-		if len(provisioningBMHs) != 2 {
-			return errors.New("Waiting for 2 BMHs to be Provisioned")
-		}
-		return nil
+		g.Expect(provisioningBMHs).To(HaveLen(2))
 	}, e2eConfig.GetIntervals(specName, "wait-machine-remediation")...).Should(Succeed())
 
 	By("Scaling up machine deployment to 3 replicas")
 	scaleMachineDeployment(ctx, bootstrapClient, clusterName, namespace, 3)
 
 	By("Waiting for one BMH to start provisioning")
-	Eventually(func(g Gomega) error {
+	Eventually(func(g Gomega) {
 		bmhs, err := getAllBmhs(ctx, bootstrapClient, namespace, specName)
-		if err != nil {
-			Logf("Error: %v", err)
-			return err
-		}
+		g.Expect(err).NotTo(HaveOccurred())
 		g.Expect(filterBmhsByProvisioningState(bmhs, bmov1alpha1.StateProvisioning)).To(HaveLen(1))
-		return nil
-	}, e2eConfig.GetIntervals(specName, "wait-machine-remediation")...)
+	}, e2eConfig.GetIntervals(specName, "wait-machine-remediation")...).Should(Succeed())
 
 	Logf("Verifying that only one BMH starts provisioning")
-	Consistently(func() error {
+	Consistently(func(g Gomega) {
 		bmhs, err := getAllBmhs(ctx, bootstrapClient, namespace, specName)
-		if err != nil {
-			Logf("Error: %v", err)
-			return err
-		}
-		Expect(filterBmhsByProvisioningState(bmhs, bmov1alpha1.StateProvisioned)).To(HaveLen(3))
-		Expect(filterBmhsByProvisioningState(bmhs, bmov1alpha1.StateProvisioning)).To(HaveLen(1))
-		return nil
-	}, e2eConfig.GetIntervals(specName, "monitor-provisioning")...)
+		g.Expect(err).NotTo(HaveOccurred())
+		g.Expect(filterBmhsByProvisioningState(bmhs, bmov1alpha1.StateProvisioned)).To(HaveLen(3))
+		g.Expect(filterBmhsByProvisioningState(bmhs, bmov1alpha1.StateProvisioning)).To(HaveLen(1))
+	}, e2eConfig.GetIntervals(specName, "monitor-provisioning")...).Should(Succeed())
```

Comment:
Same here actually. And this one is showing errors now :grimacing:

Reply:
Ah I get it now. We had wrong expectation actually. In unhealthy annotation test we start with 1 KCP and 1 worker. We mark the worker BMH as unhealthy and delete the Machine. Then wait for another BMH to be provisioned instead. So we have 1 KCP, 1 worker, 1 unhealthy and 1 available at this point.

Now we scale up the MachineDeployment to 3. Since 1 is unhealthy we only expect 1 to start provisioning. At this point we should have 2 provisioned (KCP + old worker), 1 provisioning (new worker) and 1 stuck in unhealthy/available. But the Expect had **3 provisioned**. I changed it to 2 now.

Reply:
I hope it is fixed now.
1. Wait for 1 BMH in provisioning
2. Wait for 3 BMH in provisioned
3. Check that 3 BMH stay in provisioned and 0 in provisioning, to verify that the unhealthy BMH is never provisioned.